src/ripple/app/ledger/TxHold.cpp

```diff
@@ -0,0 +1,727 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/ledger/TxHold.h>
+#include <ripple/app/tx/TransactionEngine.h>
+#include <ripple/app/ledger/LedgerConsensus.h>
+#include <numeric>
+
+namespace ripple {
+
+class CandidateTxn
+{
+    friend class TxQImpl;
+
+protected:
+    boost::intrusive::set_member_hook<> byFee_;
+
+private:
+    STTx::pointer txn_;
+
+     uint64_t      feeLevel_;
+     uint32_t      lastRelayed_;
+     uint256       txID_;
+     boost::optional<uint256> priorTxID_;
+     Account       account_;
+     boost::optional<uint32_t> lastValid_;
+     uint32_t      sequence_;
+     TransactionEngineParams transactionParams_;
+
+public:
+    CandidateTxn (STTx::ref, TransactionEngine& engine,
+        std::uint32_t loadBase, TransactionEngineParams params);
+
+    STTx::pointer getTransaction()
+    {
+        return txn_;
+    }
+
+    uint64_t getFeeLevel() const
+    {
+        return feeLevel_;
+    }
+
+    uint32_t getSequence() const
+    {
+        return sequence_;
+    }
+};
+
+CandidateTxn::CandidateTxn(
+    STTx::ref txn,
+    TransactionEngine& engine,
+    std::uint32_t loadBase,
+    TransactionEngineParams params)
+    : txn_(txn)
+    , lastRelayed_(0), txID_(txn->getTransactionID())
+    , account_(txn->getSourceAccount().getAccountID())
+    , sequence_(txn->getSequence())
+    , transactionParams_(params)
+{
+    feeLevel_ = txn->getFeeLevelPaid(loadBase, engine.getLedger()->getBaseFee());
```

Comment:
I think this `feeLevel_` initialization could be moved into the initialization list.